driver-core/src/test/functional/com/mongodb/client/model/expressions/TypeExpressionsFunctionalTest.java

```diff
@@ -65,6 +65,27 @@ public void isNumberOrTest() {
         assertExpression(99, ofNull().isNumberOr(of(99)));
     }
 
+    @Test
+    public void isIntegerOr() {
+        assertExpression(
+                1,
+                of(1).isIntegerOr(of(99)),
+                "{'$cond': [{'$and': [{'$isNumber': [1]}, {'$eq': [1, {'$round': 1}]}]}, 1, 99]}");
+        // other numeric values:
+        assertExpression(1L, of(1L).isIntegerOr(of(99)));
+        assertExpression(1.0, of(1.0).isIntegerOr(of(99)));
+        assertExpression(Decimal128.parse("1"), of(Decimal128.parse("1")).isIntegerOr(of(99)));
+        // non-numeric:
+        assertExpression(99, ofIntegerArray(1).isIntegerOr(of(99)));
+        assertExpression(99, ofNull().isIntegerOr(of(99)));
+
+        assertExpression(99,
+                of(BsonDocument.parse("{'a': 'str'}")).getField("a").isIntegerOr(of(99)));
+        // the following leads to "$round only supports numeric types, not string"
+        // because the literal type is known.
+        // assertExpression(99, of("str").isIntegerOr(of(99)));
```

Comment:
So `$and` there is no short-circuiting in MongoDB that would made this work. This behavior violates the documentation for `Expression.isIntegerOr` added in https://github.com/mongodb/mongo-java-driver/pull/1059. We can't even say "the user asserts that the value is number" here because it is an unexpected requirement given methods like `isBooleanOr`,`isDateOr`, etc.

Currently I see the following ways we could attempt to solve the problem:

1. Attempt to somehow restructure the produced MQL such that MongoDB does not try to `$round` value that is not a number (this is assuming it errors not as a result of pure validating of the MQL and the input data).
2. Attempt to use `$convert` to convert before rounding, and then `$eq` the original and the converted&rounded value.

Since `Expression.isIntegerOr` was not added in this PR, I can move this comment to `Expressions`. Either way, this seems like an important thing to address with regard to `isIntegerOr`.

Reply:
`and`, `or`, `cond` (and I think others) avoid evaluating the right branch, unless necessary. However, this does not apply if you provide certain literal values. This is bad, because it should know that the branches are not evaluated. But it is good, because the user should probably not be sending certain non-evaluated expressions.

So this only applies if the user is creating expressions that they (or we) should probably be evaluating locally. And any existing code will already work this way. I do not think we should attempt to convert all values, since this is not a problem for all values.

Reply:
Used switch, since unlike cond it avoids this problem.

Reply:
Shouldn't we use a similar approach in `isInteger`:

```java
public BooleanExpression isInteger() {
    return switchOn(on -> on
            .isNumber(v -> v.round().eq(v))
            .defaults(v -> of(false)));
}
```

to avoid similar problems:

```java
Expression yearManufacturedOrGarbage = of("");
yearManufacturedOrGarbage.switchOn(on -> on
        .isDate(v -> v.year(of("UTC")))
        // this causes an error at the moment, but works if `isNumber` is changed
        .isInteger(v -> v)
        .defaults(v -> of(0))));
```

Reply:
Yes - fixed

Reply:
We also need a test for `isInteger` covering this. One was added for `isIntegerOr` in https://github.com/mongodb/mongo-java-driver/pull/1055/commits/8e0ad99c443c48a73691cfc3206561f45c92e54c for a similar change.

Reply:
Updated. The prior test checked for this.